transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java

```diff
@@ -276,8 +277,19 @@ public boolean remove() {
         removeEntry(e);
 
         if (!e.cancelled) {
-            // only release message, notify and decrement if it was not canceled before.
-            ReferenceCountUtil.safeRelease(msg);
+            // this save both checking against the ReferenceCounted interface
+            // and makes better use of virtual calls vs interface ones
+            if (msg instanceof AbstractReferenceCountedByteBuf) {
+                try {
+                    // release now as it is flushed.
+                    ((AbstractReferenceCountedByteBuf) msg).release();
+                } catch (Throwable t) {
+                    logger.warn("Failed to release a ByteBuf: {}", msg, t);
+                }
+            } else {
+                // only release message, notify and decrement if it was not canceled before.
```

Comment:
this comment should be moved as it applies for both cases

Reply:
You meant the log one?

Reply:
No... `// only release message, notify and decrement if it was not canceled before.` Just move it above the `if (!e.cancelled) {`